Ten Elephants/Ten Elephants/Models/Meals.swift

```diff
@@ -0,0 +1,116 @@
+    let youTubeLink: String?       // strYoutube
+    let tags: [String]?            // strTags
+
+    let ingredients: [Ingredient]? // Ingredient(strIngredient, strMeasure)
+
+    enum RequiredCodingKeys: String, CodingKey, CaseIterable {
+        case strMeal,
+             idMeal,
+             strMealThumb
+    }
+
+    enum OptionalCodingKeys: String, CodingKey, CaseIterable {
+        case strCategory,
+             strArea,
+             strInstructions,
+             strYoutube,
+             strTags
+    }
+
+    init(from decoder: Decoder) throws {
+        let commonContainer = try decoder.container(keyedBy: RequiredCodingKeys.self)
+
+        // these values should always be present:
+        self.name           = try commonContainer.decode(String.self, forKey: .strMeal)
+        self.id             = try commonContainer.decode(String.self, forKey: .idMeal)
+        self.thumbnailLink  = try commonContainer.decode(String.self, forKey: .strMealThumb)
+
+        let customContainer = try decoder.container(keyedBy: OptionalCodingKeys.self)
+
+        // these values can be missing in preview model
+        self.category       = try customContainer.decodeIfPresent(String.self, forKey: .strCategory)
+        self.area           = try customContainer.decodeIfPresent(String.self, forKey: .strArea)
+        self.instructions   = try customContainer.decodeIfPresent(String.self, forKey: .strInstructions)
+        self.youTubeLink    = try customContainer.decodeIfPresent(String.self, forKey: .strYoutube)
+
+        if let nonSplitTags = try customContainer.decode(String?.self, forKey: .strTags) {
+            self.tags = nonSplitTags.components(separatedBy: ",")
+        } else {
+            self.tags = nil
+        }
+
+        var localIngredients: [Ingredient] = []
+
+        let otherContainer = try decoder.container(keyedBy: GenericCodingKeys.self)
+        let decodedKeysCommon = RequiredCodingKeys.allCases.map({ $0.rawValue })
+        let decodedKeysCustom = OptionalCodingKeys.allCases.map({ $0.rawValue })
+
+        let filteredKeys = otherContainer.allKeys.filter({
+            !decodedKeysCommon.contains($0.stringValue) && !decodedKeysCustom.contains($0.stringValue)
+        })
+
+        // assuming return when having only RequiredCodingKeys
+        guard !filteredKeys.isEmpty else {
+            self.ingredients = nil
+            return
+        }
+
+        let filteredKeysNames = filteredKeys.map({ $0.stringValue })
+
+        for num in 1...20 {
+            let ingredientKeyName = "strIndgredient\(num)"
+            let measureKeyName    = "strMeasure\(num)"
+            guard filteredKeysNames.contains(ingredientKeyName) &&
+                  filteredKeysNames.contains(measureKeyName) else {
+                self.ingredients = nil
+                assertionFailure("Unexpected data format")
+                return
+            }
+            if let ingredientKey = GenericCodingKeys(stringValue: ingredientKeyName) {
```

Comment:
Можно убрать одну вложенность
```suggestion
            if let ingredientKey = GenericCodingKeys(stringValue: ingredientKeyName),
               let measureKey = GenericCodingKeys(stringValue: measureKeyName) {
```